tools/generate.py

```diff
@@ -0,0 +1,89 @@
+#####################################################################################
+# The MIT License (MIT)
+#
+# Copyright (c) 2015-2023 Advanced Micro Devices, Inc. All rights reserved.
+#
+# Permission is hereby granted, free of charge, to any person obtaining a copy
+# of this software and associated documentation files (the "Software"), to deal
+# in the Software without restriction, including without limitation the rights
+# to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+# copies of the Software, and to permit persons to whom the Software is
+# furnished to do so, subject to the following conditions:
+#
+# The above copyright notice and this permission notice shall be included in
+# all copies or substantial portions of the Software.
+#
+# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.  IN NO EVENT SHALL THE
+# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+# THE SOFTWARE.
+#####################################################################################
+import os, sys, argparse, subprocess, te, api
+from pathlib import Path
+
+clang_format_path = Path('clang-format.exe' if os.name ==
+                         'nt' else '/opt/rocm/llvm/bin/clang-format')
+work_dir = Path().cwd()
+src_dir = (work_dir / '../src').absolute()
+migraphx_py_path = src_dir / 'api/migraphx.py'
+
+
+def clang_format(buffer, **kwargs):
+    return subprocess.run(f'{clang_format_path} -style=file',
+                          capture_output=True,
+                          shell=True,
+                          check=True,
+                          input=buffer.encode('utf-8'),
+                          cwd=work_dir,
+                          **kwargs).stdout.decode('utf-8')
+
+
+def api_generate(input_path: Path, output_path: Path):
+    with open(output_path, 'w') as f:
+        f.write(clang_format(api.invoke(input_path)))
```

Comment:
Just call the `run` function directly: `f.write(clang_format(api.invoke([migraphx_py_path, input_path])))` then there is no need to call `register_functions`.

Reply:
I tried that. It fails with an exception after the second call, saying `"updated() not being called"`. The second call doubles the functions already on the `functions` list, and `@once` attached to `process_functions()` prevents them from being updated. When I removed the decorator/reset the `had_run` flag, I ended up doubling everything: functions, parameters, structures, etc., in the resulting files. So, either I register functions once or drop everything after each call to `api.invoke()`. I decided to register functions once.